In [1]:
import pandas as pd
import numpy as np

- **injury_play.csv** can be obtained from: https://www.kaggle.com/jdemeo/analysis-concussions

In [2]:
# Import data related to uncalled penalties
injury_df = pd.read_csv('data/injury_play.csv')

# Isolate only plays with an uncalled penalty 
injury_df = injury_df[injury_df['uncalled_penalty'] == 1]
injury_df

GameKey  PlayID  GSISID Primary_Impact_Type  Primary_Partner_GSISID  \
4        54    1045   32444      Helmet-to-body                 31756.0   
7       149    3663   28128      Helmet-to-body                 29629.0   
9       218    3468   28987    Helmet-to-ground                     NaN   
10      231    1976   32214    Helmet-to-helmet                 32807.0   
16      281    1526   28987      Helmet-to-body                 30789.0   
20      364    2489   31313    Helmet-to-helmet                 32851.0   
21      364    2764   32323    Helmet-to-helmet                 31930.0   
23      392    1088   32615      Helmet-to-body                 31999.0   
27      448    2792   33838      Helmet-to-body                 31317.0   

                                      PlayDescription  \
4   (9:57) J.Berry punts 45 yards to CAR 31, Cente...   
7   (1:07) T.Morstead punts 54 yards to CAR 39, Ce...   
9   (8:03) (Punt formation) B.Kern punts 46 yards ...   
10  (1:15) (Punt formation) K.Huber punts 58 yards...   
16  (1:52) B.Kern punts 38 yards to TEN 48, Center...   
20  (11:50) J.Vogel punts 44 yards to WAS 32, Cent...   
21  (8:23) J.Vogel punts 43 yards to WAS 48, Cente...   
23  (9:57) B.Kern punts 61 yards to KC 24, Center-...   
27  (4:19) (Punt formation) M.Haack punts 42 yards...   

                                 PREVIEW LINK (5000K) Number Position  Role  \
4   http://a.video.nfl.com//films/vodzilla/153238/...     41      OLB   PRT   
7   http://a.video.nfl.com//films/vodzilla/153240/...     42       SS   PFB   
9   http://a.video.nfl.com//films/vodzilla/153243/...     89       TE   PLG   
10  https://nfl-vod.cdn.anvato.net/league/5691/18/...     81       TE   PLG   
16  https://nfl-vod.cdn.anvato.net/league/5691/18/...     89       TE   PLG   
20  http://a.video.nfl.com//films/vodzilla/153252/...     58      MLB   PLT   
21  http://a.video.nfl.com//films/vodzilla/153253/...    49d      ILB   PLG   
23  http://a.video.nfl.com//films/vodzilla/153258/...     49       LS   PLS   
27  http://a.video.nfl.com//films/vodzilla/153272/...     33       RB  PDR1   

    after_ball_punted  after_punt_received  fake  uncalled_penalty  \
4                   1                    1     0                 1   
7                   1                    1     0                 1   
9                   1                    0     0                 1   
10                  1                    1     0                 1   
16                  1                    1     0                 1   
20                  1                    1     0                 1   
21                  1                    1     0                 1   
23                  1                    1     0                 1   
27                  1                    1     0                 1   

    helmet_rule  
4             0  
7             0  
9             0  
10            1  
16            2  
20            1  
21            0  
23            0  
27            0

## Overview
- This notebook is used to point out where uncalled penalties are occuring and to back a proposal for stricter enforcement of penalties on the punt. 
- **This analysis was empirically derived.*
- **All uncalled penalties involve a return team player taking an illegal action against a punt team player and always toward their own endline**, which may indicate that the punt team should be afforded extra protection on the play (extension of the defeneseless player status).

### Proposed rule changes include:
1. **Allowance of video review of plays where an injury occurs and to allow for subsequent calls of penalties directly related to the injury to be called. If injury isn't apparent immediately and a penalty is thus an unsuitable deterant, a fine may be incurred on the perpetrater of the penalty after a confirmation later by video review (potential overlap with Rule 17 Section 2 Article 3).**
2. **Expanding the defenseless player status to punt team players who are within 5 yards of the PR and are running towards their own endzone (this would be similar to the afforded protection to the punter, Rule 12 Section 2 Article 7-a-8)**

#### Impact to the game:
- Video replay would likely be played repeatedly during times of injuries, thus there shouldn't be much time lost if a referee is making a video review during that time. A resulting penalty would impact where the subsequent play will take place so certain returns may be negated.
- Extending defenseless player status to punt team players who are within 5 yards of the PR and are running towards their own endzone would help to prevent these vicious blindside blocks. Behaviorally this would result in return team players having to shield the PR from behind by running with the PR and away from their endzone if they want to impede the pursuit of a nearby punt team player. Note that this rule would most often take affect when a punt team player is in pursuit from behind the PR (the whole aspect of running towards their own endzone).

- Just as an aside, it would be helpful to have NGS data on the referees to understand their positioning during such plays (x-y plane and facing direction) because when a penalty is not called its either the referee sees the encounter and flags or doesn't flag it, or they're not seeing the penalty at all, which compromises player safety.

## Annotated Uncalled Penalties
- NFL Rulebook: https://operations.nfl.com/media/3277/2018-nfl-rulebook_final-version.pdf
- My interpretation of the following rules may be incorrect as I am not an expert official of the game, but they look pretty darn close to what would be considered illegal.
- **Remember safety first, then teamwork.**

### Blindside Block/Tackle on Defensless Player (Rule 12, Section 2, Article 7-a-10)
- States Players in a Defenseless posture includes: A player who receives a “blindside” block when the path of the blocker is toward or parallel to his own end line.
- (Section 1, Article-1-q) is mislabeled when referencing the section on blindside blocks as a headsup
- https://operations.nfl.com/the-rules/nfl-video-rulebook/defenseless-player/
    - See time stamp for video in link: 2:21 to 2:36
        - The video states: When a blocker is moving parallel or towards his own endzone, the **legal strike zone** is below the neck and above the waist
- All penalties are by the return team and are towards their own endzone

#### Cases (GameKey, PlayID):

**(54, 1045)**: CAR player strikes above the neck (you can hear the head contact) on PIT #41 at the CAR 37; results in PIT #41 subsequently hitting head on another players knee (watch 0:07 - 0:11)     
<div align="middle">
    <video  width="720" height="540" controls src="http://a.video.nfl.com//films/vodzilla/153238/Punt_Return_by_Damiere_Byrd-IX9zynRU-20181119_154215217_5000k.mp4">
    </video>
</div>

**(392, 1088)**: BAL #42 strikes above the neck on CIN player at the BAL 20; results in transitive hit on CIN #81 (concussed player) in the head (watch 0:10 - 0:14)     
<div align="middle">
    <video  width="720" height="540" controls src="https://nfl-vod.cdn.anvato.net/league/5691/18/11/25/284954/284954_75F12432BA90408C92660A696C1A12C8_181125_284954_huber_punt_3200.mp4">
    </video>
</div>

**(364, 2489)**: WAS #45 strikes above the neck on GB #58 at the WAS 38; results in GB #58 slight tumble and then repursuit of the PR (watch 0:05 - 0:11)     
<div align="middle">
    <video  width="720" height="540" controls src="http://a.video.nfl.com//films/vodzilla/153252/44_yard_Punt_by_Justin_Vogel-n7U6IS6I-20181119_161556468_5000k.mp4">
    </video>
</div>

**(392, 1088)**: KC players strikes above the neck on TEN #49 at KC 38; results in sealing the edge for PR (watch 0:12 - 0:17)     
<div align="middle">
    <video  width="720" height="540" controls src="http://a.video.nfl.com//films/vodzilla/153258/61_yard_Punt_by_Brett_Kern-g8sqyGTz-20181119_162413664_5000k.mp4">
    </video>
</div>

**(448, 2792)**: NO #33 runs body and head into MIA players shoulder at NO 40; results in shoving MIA player out of bounds (watch 0:09 - 0:13)
- This play is notable because the blindside block (possibly illegal block above the waist) results in the blocker receiving a concussion rather than the blocker causing another individuals concussion.
<div align="middle">
    <video  width="720" height="540" controls src="http://a.video.nfl.com//films/vodzilla/153272/Haack_42_yard_punt-iP6aZSRU-20181119_165050694_5000k.mp4">
    </video>
</div>

### Illegal Block Above the Waist (Rule 12, Section 1, Article 3-b)
- States it is a foul if an offensive blocker: Blocks an opponent (from behind) in the back above the opponent’s waist, or uses his hands or arms to push an opponent from behind in a manner that affects his movement, except in close-line play. 
- https://operations.nfl.com/the-rules/nfl-video-rulebook/illegal-block-above-the-waist/
    - Video Definition: A player may not block an opponent in the back above the waist or use his hands or arms to push an opponent from behind in a way that affects his movement
- All penalties are made by the return team
    
#### Cases (GameKey, PlayID):

**(149, 3663)**: block from behind by CAR #56 on NO #42 at the CAR 47; results in tumble into two other players (watch 0:08 - 0:12)     
<div align="middle">
    <video  width="720" height="540" controls src="http://a.video.nfl.com//films/vodzilla/153240/Punt_by_Thomas_Morstead-eZpDKgMR-20181119_154525222_5000k.mp4">
    </video>
</div>

**(218, 3468)**: block from behind by IND player on TEN #89 at the TEN 35; results in player hitting head against the ground and tumbling to the TEN 42(watch 0:02 - 0:06)     
<div align="middle">
    <video  width="720" height="540" controls src="http://a.video.nfl.com//films/vodzilla/153243/Punt_by_Brett_Kern-p3udGBnb-20181119_15513915_5000k.mp4">
    </video>
</div>

**(281, 1526)**: block from behind by KC #51 on TEN #89 at the TEN 40; results in head first dive into play pile (watch 13:52 - 14:03)
- Incorrect film link was provided in dataset, correct film (https://www.youtube.com/watch?v=NsBDbLWcLyM)

**(364, 2764)**: block from behind (possible H2H from behind) by WAS player on GB #49 at the GB 42; results in head first dive into the turf (watch 0:08 - 0:11)     
<div align="middle">
    <video  width="720" height="540" controls src="http://a.video.nfl.com//films/vodzilla/153253/Justin_Vogel_2-uaXi4twT-20181119_161626398_5000k.mp4">
    </video>
</div>

# Links to other notebooks:
- Concussion play analysis with proposed rule changes: https://www.kaggle.com/jdemeo/analysis-concussions
- Analysis of punt returns: https://www.kaggle.com/jdemeo/analysis-punt-returns
- Analysis of fair catches: https://www.kaggle.com/jdemeo/analysis-fair-catches
- Preprocessing of Play Information: https://www.kaggle.com/jdemeo/preprocessing-punt-play
- Preprocessing of NGS data for the above notebooks: https://www.kaggle.com/jdemeo/preprocessing-ngs